File: leetcode/201-300/257-binary-tree-paths/bench/binary_tree_paths.c
// Benchmark workload for LeetCode #257 — Binary Tree Paths (C mirror).
// Mirrors binary_tree_paths.kara algorithm-for-algorithm.
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct { long val, left, right; } Node;

static Node *nodes; static long ncount;
static char **out; static long ocount, ocap;

static void opush(char *s) {
    if (ocount == ocap) { ocap = ocap ? ocap * 2 : 1024; out = realloc(out, ocap * sizeof(char *)); }
    out[ocount++] = s;
}

static void walk(long node, const char *prefix, long plen) {
    long left = nodes[node].left, right = nodes[node].right;
    if (left == -1 && right == -1) {
        char *d = malloc(plen + 1);
        memcpy(d, prefix, plen + 1);
        opush(d);
        return;
    }
    if (left != -1) {
        char buf[32];
        int m = snprintf(buf, sizeof buf, "->%ld", nodes[left].val);
        char *next = malloc(plen + m + 1);
        memcpy(next, prefix, plen);
        memcpy(next + plen, buf, m + 1);
        walk(left, next, plen + m);
        free(next);
    }
    if (right != -1) {
        char buf[32];
        int m = snprintf(buf, sizeof buf, "->%ld", nodes[right].val);
        char *next = malloc(plen + m + 1);
        memcpy(next, prefix, plen);
        memcpy(next + plen, buf, m + 1);
        walk(right, next, plen + m);
        free(next);
    }
}

int main(void) {
    long n = 150000, rounds = 5;
    nodes = malloc(n * sizeof(Node));
    long *open = malloc(n * sizeof(long)); long ocnt = 0;
    long state = 257257;

    state = (state * 1103515245L + 12345L) & 2147483647L;
    nodes[0].val = (state / 65536L) % 100L - 50L; nodes[0].left = -1; nodes[0].right = -1;
    ncount = 1; open[ocnt++] = 0;

    while (ncount < n) {
        state = (state * 1103515245L + 12345L) & 2147483647L;
        long wd0 = state / 65536L;
        state = (state * 1103515245L + 12345L) & 2147483647L;
        long pick = (wd0 * 32768L + state / 65536L) % ocnt;
        long parent = open[pick];
        state = (state * 1103515245L + 12345L) & 2147483647L;
        nodes[ncount].val = (state / 65536L) % 100L - 50L;
        nodes[ncount].left = -1; nodes[ncount].right = -1;
        long child = ncount++;
        if (nodes[parent].left == -1) nodes[parent].left = child;
        else { nodes[parent].right = child; open[pick] = open[ocnt - 1]; ocnt--; }
        open[ocnt++] = child;
    }

    long sink = 0;
    for (long r = 0; r < rounds; r++) {
        ocount = 0;
        char root[32];
        int rl = snprintf(root, sizeof root, "%ld", nodes[0].val);
        walk(0, root, rl);

        long h = 1;
        for (long i = 0; i < ocount; i++) {
            for (const unsigned char *p = (const unsigned char *)out[i]; *p; p++)
                h = (h * 1000003L + (long)*p) % 1000000007L;
            h = (h * 31L + 7L) % 1000000007L;
            free(out[i]);
        }
        sink = (sink * 131L + h) % 1000000007L;
    }
    printf("%ld\n", sink);
    free(nodes); free(open); free(out);
    return 0;
}

File: leetcode/201-300/257-binary-tree-paths/bench/binary_tree_paths.rs
// Benchmark workload for LeetCode #257 — Binary Tree Paths (Rust mirror).
// Mirrors binary_tree_paths.kara algorithm-for-algorithm: same LCG, same bushy
// tree construction, same string-extending DFS, same positional digest.

struct Node { val: i64, left: i64, right: i64 }

fn walk(nodes: &Vec<Node>, node: usize, prefix: &str, out: &mut Vec<String>) {
    let (left, right) = (nodes[node].left, nodes[node].right);
    if left == -1 && right == -1 {
        out.push(prefix.to_string());
        return;
    }
    if left != -1 {
        let mut next = String::with_capacity(prefix.len() + 8);
        next.push_str(prefix);
        next.push_str("->");
        next.push_str(&format!("{}", nodes[left as usize].val));
        walk(nodes, left as usize, &next, out);
    }
    if right != -1 {
        let mut next = String::with_capacity(prefix.len() + 8);
        next.push_str(prefix);
        next.push_str("->");
        next.push_str(&format!("{}", nodes[right as usize].val));
        walk(nodes, right as usize, &next, out);
    }
}

fn main() {
    let n: i64 = 150000;
    let rounds: i64 = 5;

    let mut nodes: Vec<Node> = Vec::new();
    let mut open: Vec<i64> = Vec::new();
    let mut state: i64 = 257257;
    state = (state * 1103515245 + 12345) & 2147483647;
    nodes.push(Node { val: (state / 65536) % 100 - 50, left: -1, right: -1 });
    open.push(0);

    let mut made: i64 = 1;
    while made < n {
        state = (state * 1103515245 + 12345) & 2147483647;
        let wd0 = state / 65536;
        state = (state * 1103515245 + 12345) & 2147483647;
        let pick = ((wd0 * 32768 + state / 65536) % open.len() as i64) as usize;
        let parent = open[pick] as usize;
        state = (state * 1103515245 + 12345) & 2147483647;
        nodes.push(Node { val: (state / 65536) % 100 - 50, left: -1, right: -1 });
        let child = nodes.len() as i64 - 1;
        if nodes[parent].left == -1 {
            nodes[parent].left = child;
        } else {
            nodes[parent].right = child;
            let last = open[open.len() - 1];
            open[pick] = last;
            open.pop();
        }
        open.push(child);
        made += 1;
    }

    let mut sink: i64 = 0;
    for _ in 0..rounds {
        let mut out: Vec<String> = Vec::new();
        let root_s = format!("{}", nodes[0].val);
        walk(&nodes, 0, &root_s, &mut out);

        let mut h: i64 = 1;
        for s in &out {
            for &b in s.as_bytes() {
                h = (h * 1000003 + b as i64) % 1000000007;
            }
            h = (h * 31 + 7) % 1000000007;
        }
        sink = (sink * 131 + h) % 1000000007;
    }
    println!("{}", sink);
}
